### PR TITLE
ARROW-267: [C++] Implement file format layout for IPC/RPC

### PR DESCRIPTION
**Pull request by wesm**
Standing up the PR to get some feedback. I still have to implement the read path for record batches and then add a test suite. I'd also like to add some documentation about the structure of the file format and some of the implicit assumptions (e.g. word alignment) -- I put a placeholder `IPC.md` document here for this.

I also conformed the language re: record batches (had been using "row batch" in the C++ code) to make things more sane. 

Note we are not yet able to write OS files here, see ARROW-293. Will tackle that in a follow up PR, and then we should be in a position to integration test.

### CODE REVIEW

File: cpp/src/arrow/ipc/file.cc [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+  DCHECK(position_ % 8 == 0) << "ipc::WriteRecordBatch did not perform aligned writes";
+
+  // We can infer the metadata length and length of the record batch body (the
+  // concatenated buffers) from the heade offset and the new output stream
+  // position
+  int32_t metadata_length = position_ - header_offset;
+  int32_t body_length = position_ - offset - metadata_length;
+
+  // Append metadata, to be written in the footer latera
```
Comment (julienledem) [inline cpp/src/arrow/ipc/file.cc]: s/latera/later?

File: cpp/src/arrow/ipc/file.cc [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  int64_t offset = position_;
+
+  int64_t header_offset;
+  RETURN_NOT_OK(arrow::ipc::WriteRecordBatch(columns, num_rows, sink_, &header_offset));
+  RETURN_NOT_OK(UpdatePosition());
+
+  DCHECK(position_ % 8 == 0) << "ipc::WriteRecordBatch did not perform aligned writes";
+
+  // We can infer the metadata length and length of the record batch body (the
+  // concatenated buffers) from the heade offset and the new output stream
```
Comment (julienledem) [inline cpp/src/arrow/ipc/file.cc]: s/heade/header?

**Review by julienledem — COMMENTED** [review]
overall this looks fine to me.

File: cpp/src/arrow/ipc/file.cc (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/ipc/file.cc]: fixed these typos

**Comment by wesm** [comment]
@emkornfield I had to rejigger things slightly in the IPC writer to accommodate the file format (for example: writing the size of the data header *after* the serialized flatbuffer). This all will need to be documented outside of code

**Comment by wesm** [comment]
OK, I think I'm done with the first version of this. I'd like to address inconsistencies with the specification (e.g. String is still handled like List in the IPC code, no IPC for unions) in follow up patches. 

I need to do ARROW-293 so that we can write physical files on disk, then we are in a position to write our first integration tests. @julienledem what do you think?

**Comment by julienledem** [comment]
SGTM.
Did you have a JIRA for List -> String ?
+1

**Comment by wesm** [comment]
I noted this in ARROW-261. Thanks

**Pull request closed without merging**